### ISSUE TITLE
Output buffer never flushed on restart

### ISSUE THREAD

**Issue by sathieu** [mozilla-services/heka#1738]
With the following output config:
```
[ESJsonEncoder]
type = "ESJsonEncoder"
index = "logs-%{Type}-%{%Y.%m.%d}"
es_index_from_timestamp = true
type_name = "%{Type}"
fields = ["Timestamp", "Logger", "Severity", "Payload", "Pid", "Hostname", "DynamicFields"]
  [ESJsonEncoder.field_mappings]
  Timestamp = '@timestamp'
  #Uuid = 'heka_uuid'
  #Type = 'type' # inutile, car dans type_name
  Logger = 'heka_logger'
  Severity = 'syslog_severity_code'
  Payload = 'message'
  #EnvVersion = 'heka_env_version' # inutile
  Pid = 'pid'
  Hostname = 'host'

[ElasticSearchOutput]
message_matcher = "Type != 'heka.all-report' && Type != 'heka.memstat'"
encoder = "ESJsonEncoder"
server = "http://localhost:9200"
flush_interval = 50
  [ElasticSearchOutput.buffering]
  max_file_size = 268435456  # 256MiB
  max_buffer_size = 536870912  # 512MiB
  full_action = "shutdown"
```

After a `service heka restart`, the queue is not processed. It will grow untill full.

I've looked at the code without any clue.

I've seen that #1724 has not been merged into dev yet, but this is a different problem isn't it?

**Comment by sathieu** [comment]
I switched to 0.10 branch, and it seems fixed.

**Comment by sathieu** [comment]
Reopening, as this is still happening with 0.10

**Comment by sathieu** [comment]
i'm not alone: https://mail.mozilla.org/pipermail/heka/2015-September/000772.html

I've disabled buffering for now.

**Comment by sathieu** [comment]
Any news on the heka 0.10 buffering problem?

**Comment by rafrombrc** [comment]
Starting work on this fix this week.

**Comment by rafrombrc** [comment]
I've pushed a `queue_full_shutdown_fix` branch with a change (ef12c9090b04e55cda5d7fcbaf6b72089f4976d9) that I think will resolve this issue. Can you pull and build that branch to see if it works for you?

**Comment by rafrombrc** [comment]
@ioc32 This is impacting you too... any chance you can build from my branch and see if it helps?

**Comment by ioc32** [comment]
Here's the ctest results:

The following tests FAILED:
	 10 - plugins/http (Failed)
	 11 - plugins/irc (Failed)
	 16 - plugins/process (Failed)
	 20 - plugins/udp (Failed)
	 26 - mozsvc (Failed)

The full log is here [1].

Could you have a look at it or at least confirm all tests are passing in your environment?

[1] https://gist.github.com/ioc32/1e08b656db9fabda2bba

**Comment by trink** [comment]
`make clean-heka` and then `make` this will clean up the mocks

**Comment by ioc32** [comment]
Thanks for the hints @trink!

I've issued the make clean-heka and make commands after a failed ctest (in an environment started from scratch). Then, I run ctest again but it keeps failing [1].

I'm slightly confused now, I'll try and build a release package, see if that fails too.

[1] https://gist.github.com/ioc32/2e1348786f57337551fd

**Comment by ioc32** [comment]
Well, I am starting to believe this is a classic PEBKAC problem. ctest won't pass even trying to build against master.
Will have to find another environment to build and test, in case this one's not working anymore.

**Comment by sathieu** [comment]
@ioc32 : FYI I'm having ctests problems with docker.

**Comment by nathwill** [comment]
@rafrombrc yep, looks like it works! https://gist.github.com/nathwill/c58e45c1161704f911c6.

fwiw, those same specs @ioc32 and @ioc32 reported trouble with are failing for me as well on a clean git clone on Fedora 23: https://gist.github.com/nathwill/a768f6e206a3d4e25cad, but only when using go1.5.1; if i switch back to go1.4.3, tests pass successfully.

**Comment by nathwill** [comment]
by "it works" of course i mean that a build from the queue_full_shutdown_fix branch is able to resolve a back-log after being restarted following a queue-full shutdown.

speaking of which, do you think it makes sense for queue-full shutdown to exit non-zero? hoping to avoid changing our heka service unit from `Restart=on-failure` to `Restart=always`, but can totally see the argument being made for either behavior being correct (shutdown due to "err", but as instructed...)

**Comment by timurb** [comment]
My case (described here: https://mail.mozilla.org/pipermail/heka/2015-October/000855.html) although a bit different looks to be resolved too with this patch.
I've been running patched Heka for 5 days by now and the queue haven't got overfilled on disk since that time. On other boxes which are running 0.10.0b1 the files on disk grow large as usual.

**Comment by ZeusF1** [comment]
Bug still present, I was fixing it by increasing buffer_size

Error from logs, after start
2015/11/17 10:07:14 Plugin 'ElasticSearchOutput' error: can't deliver matched message: Queue is full
(10-15 more same msgs)

version 10.0b1

**Comment by timurb** [comment]
I've just checked this with HTTPOutput buffering enabled in both stock Heka 0.10.0b1 and the `queue_full_shutdown_fix` merged in it and I'm afraid it is not get fixed.

In original Heka 0.10.0b1 I see the following messages on stop (see my configs at the bottom of the comment):
```
2015-11-17_18:51:15.16143 2015/11/17 18:51:15 Decoders shutdown complete
2015-11-17_18:51:15.16147 2015/11/17 18:51:15 Stop message sent to output 'DashboardOutput'
2015-11-17_18:51:15.16151 2015/11/17 18:51:15 Stop message sent to output 'opentsdb'
2015-11-17_18:51:15.16156 2015/11/17 18:51:15 Stop message sent to output 'relay_output'
2015-11-17_18:51:15.16160 2015/11/17 18:51:15 Plugin 'DashboardOutput': stopped
2015-11-17_18:51:15.50541 2015/11/17 18:51:15 Plugin 'relay_output': stopped
2015-11-17_18:51:15.61692 2015/11/17 18:51:15 Plugin 'opentsdb' error: updating buffer cursor: can't parse queue cursor '': invalid checkpoint format
```
After than Heka freezes and I'm only able to kill it with `kill -9`. When I start it up after that the queue logs are not removed from disk.

If I set `cursor_update_count` not to 1 but to 10 I no longer see the error nor do I see it in patched version but the freeze is still there in both cases and the queue logs are not removed from disk.

Here is my config:
```
[OpentsdbEncoder]
type = "SandboxEncoder"
filename = "lua_encoders/opentsdb_batch.lua"

  [OpentsdbEncoder.config]
  ts_from_message = true
  add_hostname_if_missing = true
  type_as_prefix = false
  tag_prefix = "tags."
  tag_list = "env adx"
  skip_fields = "Timestamp timestamp Metric metric Value value ExitStatus"

[opentsdb]
type = "HttpOutput"
message_matcher = "(Type != 'heka.all-report' && Type != 'heka.sandbox-output' && Type != 'heka.sandbox-terminated') && (TRUE || Type == 'heka.memstat')"
address = "http://opentsd.mydomain.com:4242/api/put"
encoder = "OpentsdbEncoder"
use_buffering = false # or true. This param and the buffering section was added for testing
    [opentsdb.buffering]
    max_file_size = 268435456 # 256Mb
    max_buffer_size = 107374182400 # 100Gb
    full_action = "block"
    cursor_update_count = 10

[relay_output]
type = "TcpOutput"
address = "10.20.30.40:9876"
message_matcher = "(Type != 'heka.all-report' && Type != 'heka.sandbox-output' && Type != 'heka.sandbox-terminated') && (TRUE)"
```

Nevertheless I think my a bit different case to which I've referred above (https://github.com/mozilla-services/heka/issues/1738#issuecomment-156370202) still gets fixed with the patch.

**Comment by rafrombrc** [comment]
@ZeusF1 It's known that this bug exists in v0.10.0b1, the proposed fix is currently on the `queue_full_shutdown_fix` branch, which I hope to be merging to `versions/0.10` branch today or tomorrow, to go out in a v0.10.0b2 release later this week or early next week.

@timurb I'm sorry, I'm having a hard time keeping track of what you're talking about. It seems like you're describing a different issue altogether. My testing is showing that this patch is improving the reported issue, which matches the experience of @nathwill, so I feel confident that it's helping in at least some capacity. It might make sense for you to wait until this is merged and the v0.10.0b2 is released, then you can test that out and see what your behaviour is. If there's a problem, you can file a new bug, explaining from zero what behaviour you're expecting, and how that differs from what you're seeing. There's too much in here already for me to be able to track multiple concerns with this one bug.

@nathwill Yes, exiting w/ non-zero seems reasonable. I'll look into adding that to this before it gets merged.

**Comment by timurb** [comment]
Please ignore my first comment (https://github.com/mozilla-services/heka/issues/1738#issuecomment-156370202) above, it is not relevant to this issue and relates to #1790.

**Comment by rafrombrc** [comment]
Closed in #1792.

**Comment by robison** [comment]
Not sure if I should simply open a new issue on this, but I feel pretty confident that this isn't fixed, at least as of d71617716f93a6e78423ef229393c4724adbada5

I'm able to reproduce this nearly every time by issuing a SIGTERM, and then simply waiting for the cursor update error.
